Products/ZenUtils/controlplane/application.py

```diff
@@ -13,20 +13,146 @@
+            "ERROR: Could not determine the tenantID from the environment"
+        )
+    return tid
+
+
+def _search(services, params):
+    """
+    Returns a generator which produces ServiceDefinition objects.
+    """
+    if "name" in params:
+        namepat = params["name"]
+        services = (
+            svc for svc in services if fnmatch(svc.name, namepat)
+        )
+    if "tags" in params:
+        tags = set(params["tags"])
+        includes = set(t for t in tags if not t.startswith('-'))
+        excludes = set(t[1:] for t in tags if t.startswith('-'))
+        if includes:
+            services = (
+                svc for svc in services
+                if svc.tags and (set(svc.tags) & includes == includes)
+            )
+        if excludes:
+            services = (
+                svc for svc in services
+                if not svc.tags or excludes.isdisjoint(set(svc.tags))
+            )
+    return services
+
+
+# NOTE: This cache is per-Zope instance only.
+# Unfortunately, ControlCenter does _not_ identify deleted services
+# via the 'since' API (the 'getChangesSince' method call on the client).
+# This means that one Zope will have removed the deleted service(s) from
+# its cache, but the other Zopes won't pick up this change until the TTL
+# has expired on their caches and have the services reloaded.  The result
+# is that deleted services may appear and disappear from the service listing
+# if the browser client switches Zope instances during the TTL window.
+#
+# TODO: Update the ControlCenter REST API to support identification of
+# removed services and/or implement this cache using redis or memcached (or
+# something else equivalent) and have it shared among all the Zopes.
+
+class _Cache(object):
+    """Cache for ServiceDefinition objects.
+    """
+
+    def __init__(self, client, ttl=60):
+        """Initialize an instance of _Cache.
+
+        :param client: reference to the ControlCenterClient object.
+        :param ttl: Number of seconds the cache may age before cleared.
+        """
+        self._client = client
+        self._data = None
+        self._lastUpdate = 0
+        self._ttl = ttl
+
+    def _load(self):
+        """Load all the data into the cache.
+        """
+        tenant_id = getTenantId()
+        if tenant_id is None:
+            self._data = None
+            self._lastUpdate = 0
+        else:
+            self._data = self._client.queryServices(tenantID=tenant_id)
+            self._lastUpdate = time.time()
+
+    def _refresh(self):
+        """Update the cache with changes.
+        """
+        since = int((time.time() - self._lastUpdate) * 1000)
+        # No refresh if no time has elapsed since the last update
+        if since == 0:
+            return
+        changes = self._client.getChangesSince(since)
+        if changes:
+            # Changes exist, so refresh the last update time.
+            self._lastUpdate = time.time()
+        for changed_svc in changes:
```

Comment:
The `getChangesSince` call usually returns an empty list so the `for` won't execute any iterations.